File: apps/ios/Pebbles/Features/Path/Valence/ValenceStoneStyle.swift
import SwiftUI

/// Backdrop wash and ink for one polarity of valence stone.
///
/// Mirrors the roles `EmotionPalette.pebbleFrameColors` hands a real pebble:
/// the backdrop is a soft silhouette *behind* the artwork, and the ink is what
/// the artwork's lines are drawn in. Highlight is the interesting one: backdrop
/// and ink are the *same* gradient at two intensities, so the lines read as the
/// vivid edge of a soft wash rather than as a separate colour.
struct ValenceStoneStyle {
    /// Fills the silhouette behind the artwork. Never stroked — the outline a
    /// stone reads as belongs to the ink, not to the backdrop.
    let backdrop: AnyShapeStyle
    /// Tints the pebble artwork drawn inside the backdrop.
    let ink: AnyShapeStyle

    /// Selection inverts the two roles: the wash becomes the solid and the ink
    /// goes pale, so the chosen stone reads as filled in rather than merely
    /// less faded than its neighbours. It is the same treatment
    /// `EmotionPalette.pebbleFrameColors(forIntensity: 3)` gives a hero pebble
    /// on the Path — a `light` stroke over an opaque `primary` fill.
    static func style(
        for polarity: ValencePolarity,
        scheme: ColorScheme,
        isSelected: Bool = false
    ) -> ValenceStoneStyle {
        if isSelected { return selected(for: polarity) }
        switch polarity {
        case .lowlight:
            return ValenceStoneStyle(
                backdrop: AnyShapeStyle(Color.system.muted),
                ink: AnyShapeStyle(Color.system.secondary)
            )
        case .neutral:
            // `AccentSurface` already carries a low alpha, so it lands as a
            // wash behind the opaque `AccentPrimary` artwork.
            return ValenceStoneStyle(
                backdrop: AnyShapeStyle(Color.accent.surface),
                ink: AnyShapeStyle(Color.accent.primary)
            )
        case .highlight:
            return ValenceStoneStyle(backdrop: restingWash(scheme), ink: highlightInk)
        }
    }

    private static func selected(for polarity: ValencePolarity) -> ValenceStoneStyle {
        switch polarity {
        case .lowlight:
            return ValenceStoneStyle(
                backdrop: AnyShapeStyle(Color.system.secondary),
                ink: AnyShapeStyle(Color.system.background)
            )
        case .neutral:
            return ValenceStoneStyle(
                backdrop: AnyShapeStyle(Color.accent.primary),
                ink: AnyShapeStyle(Color.accent.light)
            )
        case .highlight:
            // White rather than `AccentLight`, and against a wash taken to
            // full strength: on the resting peach the artwork had almost
            // nothing to push against.
            return ValenceStoneStyle(
                backdrop: highlightSelectedWash,
                ink: AnyShapeStyle(Color.white)
            )
        }
    }

    /// Fill for the headline word naming the picked valence. Highlight carries
    /// the same mesh its stone does, so the word and the stone read as one
    /// thing. Lowlight goes darker than its stone ink: at headline size a grey
    /// word looks disabled rather than quiet.
    static func headlineInk(for polarity: ValencePolarity) -> AnyShapeStyle {
        switch polarity {
        case .lowlight:  return AnyShapeStyle(Color.system.foreground)
        case .neutral:   return AnyShapeStyle(Color.accent.primary)
        case .highlight: return highlightInk
        }
    }

    /// The unselected stone's wash.
    ///
    /// Light mode keeps the sampled gradient at low opacity: over a light page
    /// it stays a pastel. Dark mode cannot — the same gradient over black goes
    /// muddy and opaque, and the highlight stone ends up looking nothing like
    /// its neighbours, which wear flat 10%-alpha surface colours
    /// (`AccentSurface` is `AccentPrimary` at 0.10).
    ///
    /// So dark mode joins that convention rather than fighting it, using the
    /// Joy emotion category's own `surface_color` — `#A15C08` at 0.10, the same
    /// 10% rule, in a warm gold that keeps highlight distinct from neutral's
    /// rose.
    private static func restingWash(_ scheme: ColorScheme) -> AnyShapeStyle {
        scheme == .dark
            ? AnyShapeStyle(joySurface)
            : AnyShapeStyle(highlightWash.opacity(0.35))
    }

    /// Joy's `surface_color`, copied for the same reason as the gradient
    /// samples: `EmotionPaletteService` needs the network and is not loaded
    /// when this view first draws.
    private static let joySurface = Color(hex: "#A15C081A") ?? .clear

    // MARK: - The highlight gradient

    /// The highlight material, sampled from the reference gradient at each of
    /// the mesh's own control points (patch-averaged, so no single noisy pixel
    /// decides a corner). Rose at the top left, blush across the top right,
    /// gold rising from the bottom left.
    ///
    /// Every sample lands between hue 3° and 41° — the whole thing is warm.
    /// That is what makes the ink below possible: a gradient this narrow can be
    /// saturated without becoming a rainbow, which a full-hue-wheel one cannot.
    private static let washHexes = [
        "#E7928B", "#FBA78F", "#FED6C9", "#FDC0B5",
        "#EAA68F", "#F5B592", "#FDC9B6", "#FCAA9F",
        "#EFC094", "#F8CF97", "#FBB493", "#F3968C",
        "#F1CC95", "#FADB9A", "#F1B192", "#E6908B"
    ]

    /// The same gradient as ink: each sample keeps its hue and takes a fixed
    /// saturation and lightness (HSL 0.90 / 0.52), which runs gold through
    /// orange to coral. The wash is far too light to draw the artwork with — a
    /// stone inked in it disappears against the page — so the wash fills and
    /// this twin draws.
    ///
    /// Saturated on purpose, and only affordable because the hue range is
    /// narrow. Earlier versions of this gradient spanned the wheel, where the
    /// same saturation read as a clown's palette and the alternative (darkening
    /// instead) read as mud.
    private static let inkHexes = [
        "#F32716", "#F34716", "#F34C16", "#F33816",
        "#F34E16", "#F36416", "#F35116", "#F33016",
        "#F38116", "#F39616", "#F35C16", "#F32C16",
        "#F39A16", "#F3AC16", "#F35E16", "#F32316"
    ]

    private static let meshPoints: [SIMD2<Float>] = [
        [0.0, 0.0], [0.3, 0.0], [0.7, 0.0], [1.0, 0.0],
        [0.0, 0.3], [0.2, 0.4], [0.7, 0.2], [1.0, 0.3],
        [0.0, 0.7], [0.3, 0.8], [0.7, 0.6], [1.0, 0.7],
        [0.0, 1.0], [0.3, 1.0], [0.7, 1.0], [1.0, 1.0]
    ]

    /// The wash, taken up to full strength for the selected stone: the same
    /// hues at HSL 0.92 / 0.58. The resting wash is too pale to carry a white
    /// outline — white on it has almost no contrast at the gold corner — and
    /// selection is exactly where the stone needs to shout.
    private static let selectedWashHexes = [
        "#F64031", "#F65D31", "#F66231", "#F64F31",
        "#F66331", "#F67731", "#F66631", "#F64931",
        "#F69131", "#F6A331", "#F67031", "#F64531",
        "#F6A731", "#F6B731", "#F67231", "#F63C31"
    ]

    private static let highlightWash = gradient(from: washHexes)
    private static let highlightSelectedWash = gradient(from: selectedWashHexes)
    private static let highlightInk = gradient(from: inkHexes)

    /// A mesh on iOS 18, and below that a linear run through the same four
    /// diagonal samples — the same colours in the same corner order, so a 17
    /// device gets a coherent stone rather than a different-looking one.
    ///
    /// There is no gradient token in the design system, and this deliberately
    /// does not create one: promote it when a second surface needs it.
    private static func gradient(from hexes: [String]) -> AnyShapeStyle {
        let colors = hexes.map { Color(hex: $0) ?? .clear }
        if #available(iOS 18, *) {
            return AnyShapeStyle(
                MeshGradient(width: 4, height: 4, points: meshPoints, colors: colors)
            )
        }
        return AnyShapeStyle(
            LinearGradient(
                colors: [colors[0], colors[5], colors[10], colors[15]],
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )
        )
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/path/valence/ValenceStoneStyle.kt
package app.pbbls.android.features.path.valence

import android.graphics.Bitmap
import android.graphics.BitmapShader
import android.graphics.Matrix
import android.graphics.Shader
import androidx.compose.material3.ColorScheme
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ShaderBrush
import androidx.compose.ui.graphics.SolidColor
import androidx.core.graphics.toColorInt
import app.pbbls.android.core.model.ValencePolarity

/**
 * Backdrop wash and ink for one polarity of valence stone — ports iOS
 * `ValenceStoneStyle.swift`.
 *
 * Mirrors the roles `EmotionPalette.pebbleFrameColors` hands a real pebble: the
 * backdrop is a soft silhouette *behind* the artwork, and the ink is what the
 * artwork's lines are drawn in. Highlight is the interesting one: backdrop and
 * ink are the *same* gradient at two intensities, so the lines read as the
 * vivid edge of a soft wash rather than as a separate colour.
 *
 * [backdropAlpha] rides beside the brush because a `ShaderBrush` carries no
 * opacity of its own — the way iOS spells `highlightWash.opacity(0.35)`.
 */
internal data class ValenceStoneStyle(
    val backdrop: Brush,
    val ink: Brush,
    val backdropAlpha: Float = 1f,
)

/**
 * The nine stones' materials, resolved from the active [ColorScheme].
 *
 * Selection **inverts the two roles**: the wash becomes the solid and the ink
 * goes pale, so the chosen stone reads as filled in rather than merely less
 * faded than its neighbours. It is the same treatment
 * `EmotionPalette.pebbleFrameColors(intensity = 3)` gives a hero pebble on the
 * Path — a `light` stroke over an opaque `primary` fill.
 */
internal object ValenceStoneStyles {
    /**
     * Joy's `surface_color` at 10%, copied for the same reason as the gradient
     * samples: `EmotionPaletteService` needs the network and is not loaded when
     * this view first draws. Emotion data, so it is parsed from its hex like
     * the mesh rather than written as a theme literal.
     */
    private val joySurface =
        Color(ValenceMesh.JOY_SURFACE_HEX.toColorInt()).copy(alpha = 0.10f)

    private val wash = MeshShaderBrush(ValenceMesh.washHexes)
    private val selectedWash = MeshShaderBrush(ValenceMesh.selectedWashHexes)
    private val ink = MeshShaderBrush(ValenceMesh.inkHexes)

    fun style(
        polarity: ValencePolarity,
        isSelected: Boolean,
        isDark: Boolean,
        scheme: ColorScheme,
    ): ValenceStoneStyle =
        if (isSelected) {
            when (polarity) {
                // The inverse pair is M3's own "flip the surface" role, which is
                // exactly what selection does to the grey stone.
                ValencePolarity.LOWLIGHT -> {
                    ValenceStoneStyle(SolidColor(scheme.inverseSurface), SolidColor(scheme.inverseOnSurface))
                }

                ValencePolarity.NEUTRAL -> {
                    ValenceStoneStyle(SolidColor(scheme.primary), SolidColor(scheme.onPrimary))
                }

                // White rather than `onPrimary`, and against a wash taken to
                // full strength: on the resting peach the artwork had almost
                // nothing to push against. The mesh is baked from fixed samples
                // and does not follow the theme, so its ink cannot either.
                ValencePolarity.HIGHLIGHT -> {
                    ValenceStoneStyle(selectedWash, SolidColor(Color.White))
                }
            }
        } else {
            when (polarity) {
                ValencePolarity.LOWLIGHT -> {
                    ValenceStoneStyle(
                        SolidColor(scheme.surfaceContainerHighest),
                        SolidColor(scheme.onSurfaceVariant),
                    )
                }

                // `primary` rather than the paired `onPrimaryContainer`: the
                // artwork is what makes the neutral stone rose, and the two
                // tones sit far enough apart for line art (3:1).
                ValencePolarity.NEUTRAL -> {
                    ValenceStoneStyle(SolidColor(scheme.primaryContainer), SolidColor(scheme.primary))
                }

                ValencePolarity.HIGHLIGHT -> {
                    ValenceStoneStyle(restingWash(isDark), ink, restingWashAlpha(isDark))
                }
            }
        }

    /**
     * Fill for the headline word naming the picked valence. Highlight carries
     * the same gradient its stone does, so the word and the stone read as one
     * thing. Lowlight goes darker than its stone ink: at headline size a grey
     * word looks disabled rather than quiet.
     */
    fun headlineInk(
        polarity: ValencePolarity,
        scheme: ColorScheme,
    ): Brush =
        when (polarity) {
            ValencePolarity.LOWLIGHT -> SolidColor(scheme.onSurface)
            ValencePolarity.NEUTRAL -> SolidColor(scheme.primary)
            ValencePolarity.HIGHLIGHT -> ink
        }

    /**
     * The unselected highlight stone's wash.
     *
     * Light mode keeps the sampled gradient at low opacity: over a light page
     * it stays the pastel it was sampled from. Dark mode cannot — the same
     * gradient over black goes muddy and opaque, and the highlight stone ends
     * up looking nothing like its neighbours, which wear flat container
     * colours (neutral's `primaryContainer`). So dark mode joins that
     * convention rather than fighting it, in Joy's own warm gold, which keeps
     * highlight distinct from neutral's rose. That gold is emotion data (like
     * the mesh samples), not a theme role: no scheme role means "joy".
     */
    private fun restingWash(isDark: Boolean): Brush = if (isDark) SolidColor(joySurface) else wash

    private fun restingWashAlpha(isDark: Boolean): Float = if (isDark) 1f else 0.35f
}

/**
 * A baked mesh gradient as a Compose [Brush].
 *
 * The pixels come from [ValenceMesh.bake]; the shader stretches that small
 * square over whatever it fills, with linear filtering so the upscale stays
 * smooth (`FILTER_MODE_LINEAR` is API 31, well under minSdk 33). `ShaderBrush`
 * memoizes the shader per size, so a stone pays for its own shader once.
 *
 * The bitmap is built lazily and kept for the process: three gradients at
 * [ValenceMesh.RESOLUTION]² are ~48KB in total, and rebuilding one per
 * composition would be sixteen distance calculations per pixel per frame.
 */
private class MeshShaderBrush(
    private val hexes: List<String>,
) : ShaderBrush() {
    private val bitmap: Bitmap by lazy {
        Bitmap.createBitmap(
            ValenceMesh.bake(hexes),
            ValenceMesh.RESOLUTION,
            ValenceMesh.RESOLUTION,
            Bitmap.Config.ARGB_8888,
        )
    }

    // `android.graphics.Shader` rather than the Compose alias for it: they are
    // the same class, and only the platform name lets `Shader.TileMode` resolve.
    override fun createShader(size: Size): Shader =
        BitmapShader(bitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP).apply {
            setFilterMode(BitmapShader.FILTER_MODE_LINEAR)
            setLocalMatrix(
                Matrix().apply {
                    setScale(size.width / bitmap.width, size.height / bitmap.height)
                },
            )
        }
}
